### PR TITLE
Add possibility to upgrade the current session to a new TCP handler

### PR DESCRIPTION
**Pull request by mfelsche**
and add a Websocket example demonstrating this new feature.

### CODE REVIEW

File: examples/websocket_echo_server/lock.json [hunk 1]
```diff
@@ -0,0 +1,32 @@
+{
```
Comment (SeanTAllen) [inline examples/websocket_echo_server/lock.json:1]: can we not commit the lock.json? that tends to lead to wtf with tests etc.

is this required? i assume it isn't. (if we do remove, lets add it to the .gitignore)

File: http_server/request.pony [hunk 2]
```diff
@@ -8,23 +8,47 @@ primitive HTTP11 is _Version
...
+    end
+
 
 primitive HTTP10 is _Version
   """
   HTTP/1.0
   """
   fun string(): String iso^ => recover iso String(8).>append("HTTP/1.0") end
   fun to_bytes(): Array[U8] val => [as U8: 'H'; 'T'; 'T'; 'P'; '/'; '1'; '.'; '0']
+  fun u64(): U64 => 'HTTP/1.0'
```
Comment (SeanTAllen) [inline http_server/request.pony:27]: im curious what is going on here.

File: http_server/session.pony [hunk 3]
```diff
@@ -1,4 +1,5 @@
 use "valbytes"
+use "net"
```
Comment (SeanTAllen) [inline http_server/session.pony:2]: can we put this before valbytes?

"standard lib before '3rd party' packages"?

**Comment by SeanTAllen** [comment]
@mfelsche seems like we should add `github.com/mfelsche/pony-websocket` as a ponylang project. Are you up for doing that?

**Comment by SeanTAllen** [comment]
Or perhaps @mfelsche, would it make sense to incorporate `github.com/mfelsche/pony-websocket` into this library?

**Comment by SeanTAllen** [comment]
This needs release notes.

File: examples/websocket_echo_server/lock.json [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+      "revision": "main"
+    },
+    {
+      "locator": "github.com/ponylang/json.git",
+      "revision": "0.1.0"
+    },
+    {
+      "locator": ".",
+      "revision": "main"
+    },
```
Comment (SeanTAllen) [inline examples/websocket_echo_server/lock.json:14]: I'd rather we use "../" in the use in examples.

File: Makefile [hunk 5]
```diff
@@ -40,9 +40,11 @@ endif
 PONYC := $(PONYC) $(SSL)
 
 SOURCE_FILES := $(shell find $(SRC_DIR) -name *.pony)
-EXAMPLES := $(notdir $(shell find $(EXAMPLES_DIR)/* -type d))
-EXAMPLES_SOURCE_FILES := $(shell find $(EXAMPLES_DIR) -name *.pony)
+EXAMPLES := $(notdir $(shell find $(EXAMPLES_DIR)/* -maxdepth 0 -type d -not -name websocket_echo_server))
+EXAMPLES_SOURCE_FILES := $(shell find $(EXAMPLES_DIR) -name *.pony -not -path '**/websocket_echo_server/*')
```
Comment (SeanTAllen) [inline Makefile:44]: why are we excluding the websocket example from being checked if it compiles?

File: Makefile [hunk 6]
```diff
@@ -54,12 +56,17 @@ $(tests_binary): $(SOURCE_FILES) | $(BUILD_DIR)
...
 
 $(EXAMPLES_BINARIES): $(BUILD_DIR)/%: $(SOURCE_FILES) $(EXAMPLES_SOURCE_FILES) | $(BUILD_DIR)
 	$(GET_DEPENDENCIES_WITH)
 	$(PONYC) -o $(BUILD_DIR) $(EXAMPLES_DIR)/$*
 
+$(WEBSOCKET_EXAMPLES_BINARY): $(BUILD_DIR)/%: $(SOURCE_FILES) $(WEBSOCKET_EXAMPLE_SOURCE_FILES) | $(BUILD_DIR)
+	cd $(EXAMPLES_DIR)/$* && \
+		$(GET_DEPENDENCIES_WITH) && \
+		$(PONYC) -o ../../$(BUILD_DIR) $(EXAMPLES_DIR)/$*
+
```
Comment (SeanTAllen) [inline Makefile:69]: Why is this separate? having different commands to build examples seems like a footgun for people maintaining. (I'm think of a frustrated Sean in the future).

File: examples/websocket_echo_server/main.pony [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+        end
+        _session.send_raw(
+          _response_builder.set_status(StatusSwitchingProtocols)
+            .add_header("Upgrade", "websocket")
+            .add_header("Connection", "Upgrade")
+            .add_header("Sec-WebSocket-Accept", sec_ws_accept)
+            .finish_headers()
+            .build(),
+          request_id
+        )
```
Comment (jemc) [inline examples/websocket_echo_server/main.pony:79]: Should this be integrated be integrated with the `_session.upgrade` method so that this "Switching Protocols" response will definitely happen (without the user needing to remember to implement this)?

I think it would be better to make this all in one call rather than asking the protocol switching handler to construct a raw response.

**Comment by SeanTAllen** [comment]
If the websocket library isn't a ponylang project then the example shouldnt be part of this repo. The build could be broken by the websocket library falling out of date and maintainers for this package who aren't Matthias wouldn't be able to fix without removing the example.

I think either example goes or we commit to making the websocket library a ponylang project.

IF we go for "example goes", then I think we should take the example and it can be an example repo unto itself OR we can do a blog post on the website after this is released with an example (and perhaps some FAQ fun on the website).

File: http_server/session.pony [hunk 8]
```diff
@@ -243,5 +244,10 @@ trait tag Session
     """
     None
 
+  be upgrade(notify: TCPConnectionNotify iso) =>
+    """
+    Upgrade this TCP connection to another handler
+    """
```
Comment (jemc) [inline http_server/session.pony:250]: I know that "upgrade" is [the official HTTP jargon for this](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Upgrade), but it might be confusing to people at first. I recommend something like "upgrade_protocol" to make this more clear. I also recommend linking to those docs, as below:

```suggestion
  be upgrade_protocol(notify: TCPConnectionNotify iso) =>
    """
    Upgrade this TCP connection to another handler.
    See https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Upgrade
    """
```

File: examples/websocket_echo_server/lock.json (hunk 1 above)
Comment (mfelsche) [reply examples/websocket_echo_server/lock.json:1]: This is in the example. We can totally keep it out. My thought was that it is usually good practice to include it to improve reproducability of the build of this corral project, as the lock is where the version/git-tag pinning happens.

File: http_server/request.pony (hunk 2 above)
Comment (mfelsche) [reply http_server/request.pony:27]: This is a preparation for checking for the correct HTTP version part of a request by checking for a full u64 instead of comparing strings. This is a SWAR technique i saw in another HTTP parser library. The multi-char literal will create the following hex int: `0x485454502F312E30`.

File: Makefile (hunk 5 above)
Comment (mfelsche) [reply Makefile:44]: it is included below in separate variables as we need a separate `corral fetch` and `corral run` for it.

File: Makefile (hunk 6 above)
Comment (mfelsche) [reply Makefile:69]: I separated it out because it needs to fetch its own dependencies.

File: http_server/request.pony (hunk 2 above)
Comment (SeanTAllen) [reply http_server/request.pony:27]: Can you document that?

File: examples/websocket_echo_server/main.pony (hunk 7 above)
Comment (sacovo) [reply examples/websocket_echo_server/main.pony:79]: Are websockets the only reason one would upgrade the session, or might there be other reasons and the header would look different?

**Comment by SeanTAllen** [comment]
@mfelsche for the largest question first so we can move this along, do you want to pull the example out that relies on an external 3rd party library out and use some other location to detail the example or do you want to have the websockets library become a ponylang project that we make sure works with pony language changes going forward?

File: examples/websocket_echo_server/main.pony (hunk 7 above)
Comment (SeanTAllen) [reply examples/websocket_echo_server/main.pony:79]: Upgrade is a general part of the HTTP protocol, so it isn't websocket specific.

**Comment by sacovo** [comment]
Maybe remove the example for now, and if/when the websocket library is a ponylang project it can be added back in?

**Comment by mfelsche** [comment]
Sorry for the late reply. In order to move this forward, lets for now move this example out. I will improve the documentation of the `upgrade_protocol` behaviour. @sacovo Does this as is solve your use case for websocket support?

**Comment by sacovo** [comment]
Yes, it does. Thank you for your work!

File: examples/websocket_echo_server/main.pony (hunk 7 above)
Comment (mfelsche) [reply examples/websocket_echo_server/main.pony]: I thought about integrating sending a 101 Response with in the `Session.upgrade_protocol` function. I initially voted against it, because i wanted to keep the library as raw and low-level as possible. While nearly all protocol upgrades will be used for Websockets, I'd rather not block the path for upgrading to other protocols. And people might want to add special headers to their response and we have multiple API styles used in this library for sending responses, so a generic upgrade_protocol method that does all the upgrading boilerplate while still being generic enough so people can change the response as they see fit, didn't seem worthwile to me.

I'd suggest to move the discussion about a more higher-level API for this to new Issue if need be.

File: Makefile [hunk 9]
```diff
@@ -40,8 +40,8 @@ endif
 PONYC := $(PONYC) $(SSL)
 
 SOURCE_FILES := $(shell find $(SRC_DIR) -name *.pony)
-EXAMPLES := $(notdir $(shell find $(EXAMPLES_DIR)/* -type d))
-EXAMPLES_SOURCE_FILES := $(shell find $(EXAMPLES_DIR) -name *.pony)
+EXAMPLES := $(notdir $(shell find $(EXAMPLES_DIR)/* -maxdepth 0 -type d -not -name websocket_echo_server))
+EXAMPLES_SOURCE_FILES := $(shell find $(EXAMPLES_DIR) -name *.pony -not -path '**/websocket_echo_server/*')
```
Comment (SeanTAllen) [inline Makefile:44]: This should be rolled back to how it was previously.

File: http_server/_server_connection.pony [hunk 10]
```diff
@@ -1,9 +1,10 @@
-use "net"
 use "collections"
-use "valbytes"
 use "debug"
+use "net"
 use "time"
 
+use "valbytes"
```
Comment (SeanTAllen) [inline http_server/_server_connection.pony:6]: This is unrelated at this point, no new uses were added so I think this should be rolled bacl

File: http_server/_server_conn_handler.pony [hunk 11]
```diff
@@ -1,5 +1,5 @@
-use "net"
 use "debug"
+use "net"
```
Comment (SeanTAllen) [inline http_server/_server_conn_handler.pony:2]: This is unrelated at this point, no new uses were added so I think this should be rolled back

File: http_server/_test_connection_handling.pony [hunk 12]
```diff
@@ -1,5 +1,5 @@
-use "pony_test"
```
Comment (SeanTAllen) [inline http_server/_test_connection_handling.pony:1]: This is unrelated at this point, no new uses were added so I think this should be rolled back

File: http_server/_test_headers.pony [hunk 13]
```diff
@@ -1,6 +1,7 @@
 use "debug"
 use "pony_check"
```
Comment (SeanTAllen) [inline http_server/_test_headers.pony:2]: This is unrelated at this point, no new uses were added so I think this should be rolled back

File: http_server/_test_pipelining.pony [hunk 14]
```diff
@@ -1,7 +1,8 @@
 use "net"
-use "time"
 use "pony_test"
 use "random"
```
Comment (SeanTAllen) [inline http_server/_test_pipelining.pony:3]: This is unrelated at this point, no new uses were added so I think this should be rolled back

File: http_server/_test_request_parser.pony [hunk 15]
```diff
@@ -1,6 +1,7 @@
 use "debug"
 use "itertools"
 use "pony_test"
+
```
Comment (SeanTAllen) [inline http_server/_test_request_parser.pony:4]: this should be rolled back

File: http_server/handler.pony [hunk 16]
```diff
@@ -15,7 +15,7 @@ interface Handler
```
Comment (SeanTAllen) [inline http_server/handler.pony:15]: Is this change related to the upgrade? Should this be a separate PR?

File: http_server/request_parser.pony [hunk 17]
```diff
@@ -1,6 +1,7 @@
-use "valbytes"
 use "debug"
 
+use "valbytes"
+
```
Comment (SeanTAllen) [inline http_server/request_parser.pony:4]: This is unrelated at this point, no new uses were added so I think this should be rolled back

File: http_server/response.pony [hunk 18]
```diff
@@ -1,6 +1,7 @@
-use "valbytes"
 use "format"
 
+use "valbytes"
+
 interface val Response is ByteSeqIter
```
Comment (SeanTAllen) [inline http_server/response.pony:5]: This is unrelated at this point, no new uses were added so I think this should be rolled back

File: http_server/sync_handler.pony [hunk 19]
```diff
@@ -1,6 +1,7 @@
-use "valbytes"
 use "debug"
 
+use "valbytes"
+
```
Comment (SeanTAllen) [inline http_server/sync_handler.pony:4]: This is unrelated at this point, no new uses were added so I think this should be rolled back

File: .release-notes/next-release.md [hunk 20]
```diff
@@ -0,0 +1,3 @@
+## Add `Session.upgrade_protocol` behaviour
+
+This can be used to upgrade the underlying TCP connection to a new incompatible protocol, like websockets.
```
Comment (jemc) [inline .release-notes/next-release.md:3]:
```suggestion
This can be used to upgrade the underlying TCP connection to a new incompatible protocol, like websockets.

Calling this new behaviour allows this TCP connection to be upgraded to another handler, serving another protocol (e.g. [WebSocket](https://www.rfc-editor.org/rfc/rfc6455.html)).

Note that this method does not send an HTTP Response with a status of 101. This needs to be done before calling this behaviour. Also, the passed in `notify` will not have its methods [accepted](https://stdlib.ponylang.io/net-TCPConnectionNotify/#connected) or [connected](https://stdlib.ponylang.io/net-TCPConnectionNotify/#connected) called, as the connection is already established.

After calling this behaviour, this session and the connected Handler instance will not be called again, so it is necessary to do any required clean up right after this call.

See:
  - [Protocol Upgrade Mechanism](https://developer.mozilla.org/en-US/docs/Web/HTTP/Protocol_upgrade_mechanism)
  - [Upgrade Header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Upgrade)
```

**Review by SeanTAllen — APPROVED** [review]

**Review by jemc — APPROVED** [review]
